src/map.cpp

```diff
@@ -357,6 +357,9 @@ void map::vehmove()
         level_cache &cache = get_cache( zlev );
         for( vehicle *veh : cache.vehicle_list ) {
             veh->gain_moves();
+            if( !veh ) {
```

Comment:
If `veh` is a null pointer, the line above has already triggered UB. The check for null is pointless at that point. And why would the vehicle cache even contain a null pointer?

Reply:
This is due to the vehicle potentially destroying itself during gain_moves() , when the summon timer is decremented, and therefore if it is destroyed it should not be allowed to continue with the following code.

Reply:
It's still a pointless check. Either `veh` is non-null, which means the condition is false, or `veh` is null (which means it was null in the previous line as well), which would have triggered UB there.

Testing a pointer for being null *after* it has been dereferenced is too late. The damage is already done. If at all, you need to test it before using it. But this brings up the question how a null pointer got into `cache.vehicle_list`.

Reply:
A null pointer didnt get into cache.vehicle_list()
Testing the pointer before using it is what this check does.
1 line above this check, is veh.gain_moves(), in that function, it can call destroy_vehicle() then *immediately* return. 
No UB happens because no other code is run after destruction, then this here *immediately* checks if it is now non-null, this is the same thing that happens in other vehicle code areas where the vehicle destroys itself, afterwards, it checks itself.

Reply:
> Testing the pointer before using it is what this check does.

It is actually used *before* that test.

How can pointer `veh` change? It is *defined* within this function (as part of the loop), its address is never given to another function, nor is a reference to it given to another function. No code outside this loop knows about this variable. You can even make that pointer a constant and it will *still* compile:
```C++
            for( vehicle *const veh : cache.vehicle_list ) {
```


----

> 1 line above this check, is veh.gain_moves(), in that function, it can call destroy_vehicle() then immediately return.

`gain_moves` doesn't destroy any vehicle. Existing code to destroy vehicles is located in `act_on_map`, which is called from `map::vehproceed`.

----

The only *new* code that destroys a vehicle is within `decrement_summon_timer`. That function is only called in `vehicle::act_on_map`. That function in turn already contains a call to `m.destroy_vehicle`, so its callers already expect that. Your new code does not change that, you don't need more code to handle the destruction of the magic vehicle. All the surrounding code already handles the destroyed vehicle.

Reply:
Huh, I guess I moved it somehwere better, I thought I had left the decrement_summon stuff in gain_moves() , but yeah, that makes sense now, thanks.